File: haskell/Problem14.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 14
-- Longest Collatz sequence

module Problem14 where

import Common (memoize)

import Data.List (maximumBy)
import Data.Ord (comparing)

limit :: Int
limit = 999999

collatzLen :: Int -> Int
collatzLen = memoize (1, limit) $ go 1
  where
    go count 1 = count
    go count n = count + collatzLen next
      where
        next = if even n then n `div` 2 else 3 * n + 1

solve :: Int
solve = maximumBy (comparing collatzLen) [1..limit]

File: c++/src/problem_14.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 14
// Longest Collatz sequence

namespace problem_14 {

long collatz_len(long n) {
	long len = 0;
	while (n != 1) {
		if (n % 2 == 0) {
			n /= 2;
		} else {
			n = 3 * n + 1;
		}
		++len;
	}
	return len;
}

long solve() {
	long n = 0;
	long max_len = 0;
	for (long i = 1; i < 1000000; ++i) {
		const long len = collatz_len(i);
		if (len > max_len) {
			n = i;
			max_len = len;
		}
	}
	return n;
}

} // namespace problem_14
